File: webgldemo/RefFrame.js
class RefFrame {
    constructor(parent = null) {
        this.validSubtree = true;
        this.parent       = null;
        this.child        = null;
        this.next         = null;
        this.localM       = new Matrix4x4;
        this.modelM       = new Matrix4x4;

        if (parent instanceof RefFrame) {
            this.validSubtree  = false;
            this.parent        = parent;
            this.next          = parent.child;
            this.parent.child  = this;
        }
    }

    invalidateSubtree() {
        if (this.validSubtree) {
            for (let i = this.child; i !== null; i = i.next) {
                i.invalidateSubtree();
            }
            this.validSubtree = false;
            this.invalidateSubClass();
        }
    }

    validateAscending() {
        if (!this.validSubtree) {
            if (this.parent) {
                this.parent.validateAscending();
                this.modelM = this.localM.mul(this.parent.modelM);
            }
            this.validSubtree = true;
        }
    }

    get localMatrix() {
        return this.localM;
    }

    set localMatrix(matrix) {
        this.localM = matrix;
        this.invalidateSubtree();
    }

    get modelMatrix() {
        this.validateAscending();
        return this.modelM;
    }

    set modelMatrix(matrix) {
        // worldMatrix = localMatrix * parent->worldMatrix
        // worldMatrix / parent->worldMatrix = localMatrix
        this.localM = matrix.mul(this.parent.modelMatrix.inverse());
        invalidateSubtree();
    }

    translate(v, relative2) {
        if (relative2 === this || relative2 === null) { // relative to own axes
            const x = this.localM._41;
            const y = this.localM._42;
            const z = this.localM._43;
            this.localM._41 = x + (v.x * this.localM._11) + (v.y * this.localM._21) + (v.z * this.localM._31);
            this.localM._42 = y + (v.x * this.localM._12) + (v.y * this.localM._22) + (v.z * this.localM._32);
            this.localM._43 = z + (v.x * this.localM._13) + (v.y * this.localM._23) + (v.z * this.localM._33);

        } else if (relative2.parent === null) { // relative to root axes
            Vector1x4 d = relative2->mapPos(Vector1x4(v.x, v.y, v.z, 0.f), m_parent.get());
            this.localM._41 += d.m_x;
            this.localM._42 += d.m_y;
            this.localM._43 += d.m_z;

        } else if (relative2 === this.parent) { // relative to parent axes
            this.localM._41 += v.x;
            this.localM._42 += v.y;
            this.localM._43 += v.z;

        } else { // relative to arbitrary axes
            const d = relative2.mapPos(new Vector1x4(v.x, v.y, v.z, 0.0), this.parent);
            this.localM._41 += d.m_x;
            this.localM._42 += d.m_y;
            this.localM._43 += d.m_z;
        }
        this.invalidateSubtree();
    }

    rotateX(theta, relative2) {
        if (relative2 == this || relative2 == null) { // relative to own axes
            const rotx = new Matrix4x4; rotx.setRx(theta);
            this.localMatrix = rotx.mul(this.localMatrix);
        } else if (relative2 == m_parent.get()) { // relative to parent
            const rotx = new Matrix4x4; rotx.setRx(theta);
            this.localMatrix = this.localMatrix.mul(rotx);
        } else { // relative to arbitrary axes
            throw new Error('Cannot rotate relative to arbitrary axis');
        }
        invalidateSubtree();
    }

    rotateY(theta, relative2) {
        if (relative2 === this || relative2 === null) { // relative to own axes
            const roty = new Matrix4x4; roty.setRy(theta);
            this.localMatrix = roty.mul(this.localMatrix);
        } else if (relative2 == m_parent.get()) { // relative to parent
            const roty = new Matrix4x4; roty.setRy(theta);
            this.localMatrix = this.localMatrix.mul(roty);
        } else { // relative to arbitrary axes
            throw new Error('Cannot rotate relative to arbitrary axis');
        }
        invalidateSubtree();
    }

    rotateZ(theta, relative2) {
        if (relative2 === this || relative2 === null) { // relative to own axes
            const rotz = new Matrix4x4; rotz.setRz(theta);
            this.localMatrix = rotz.mul(this.localMatrix);
        } else if (relative2 == m_parent.get()) { // relative to parent
            const rotz = new Matrix4x4; rotz.setRz(theta);
            this.localMatrix = this.localMatrix.mul(rotz);
        } else { // relative to arbitrary axes
            throw new Error('Cannot rotate relative to arbitrary axis');
        }
        invalidateSubtree();
    }

    mapPos(v, tgt) {
        if (tgt == this.parent) {
            return v.mul(this.localMatrix);
        } else if (tgt.parent == null) {
            return v.mul(this.modelMatrix);
        } else {
            return v.mul(this.modelMatrix).mul(tgt.modelMatrix.inverse());
        }
    }
}

File: d3d11/RefFrame.cpp
#include "StdAfx.h"
#include "RefFrame.h"

namespace d3d11demo {

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //
    void RefFrame::invalidateSubtree() {
        if (m_validSubtree) {
            for (auto i = m_child; i != nullptr; i = i->m_next) {
                i->invalidateSubtree();
            }
            m_validSubtree = false;
            invalidateSubClass();
        }
    }

    void RefFrame::validateAscending() const {
        if (!m_validSubtree) {
            if (m_parent) {
                m_parent->validateAscending();
                m_modelMatrix = m_localMatrix * m_parent->m_modelMatrix;
            }
            m_validSubtree = true;
        }
    }

    RefFrame::RefFrame(shared_ptr<RefFrame> parent) {
        if (parent) {
            m_validSubtree  = false;
            m_parent        = parent;
            m_next          = parent->m_child;
            parent->m_child = this;
        }
    }

    RefFrame::~RefFrame() {
        if (m_parent) {
            for (auto i = m_parent->m_child; i->m_next != nullptr; i = i->m_next) {
                if (i->m_next == this) {
                    i->m_next = m_next;
                    break;
                }
            }
            if (m_parent->m_child == this) {
                m_parent->m_child = m_next;
            }
        }
    }

    const Matrix4x4& RefFrame::getLocalMatrix() const {
        return m_localMatrix;
    }

    const Matrix4x4& RefFrame::getModelMatrix() const {
        validateAscending();
        return m_modelMatrix;
    }

    void RefFrame::setLocalMatrix(const Matrix4x4& mtx) {
        m_localMatrix = mtx;
        invalidateSubtree();
    }

    void RefFrame::setModelMatrix(const Matrix4x4& mtx) {
        // worldMatrix = localMatrix * parent->worldMatrix
        // worldMatrix / parent->worldMatrix = localMatrix
        m_localMatrix = mtx * !m_parent->getModelMatrix();
        invalidateSubtree();
    }

    void RefFrame::translate(const Vector1x4& v, const RefFrame* relative2) {
        if (relative2 == this || relative2 == nullptr) { // relative to own axes
            const float x = m_localMatrix.m_41;
            const float y = m_localMatrix.m_42;
            const float z = m_localMatrix.m_43;
            m_localMatrix.m_41 = x + (v.m_x * m_localMatrix.m_11) + (v.m_y * m_localMatrix.m_21) + (v.m_z * m_localMatrix.m_31);
            m_localMatrix.m_42 = y + (v.m_x * m_localMatrix.m_12) + (v.m_y * m_localMatrix.m_22) + (v.m_z * m_localMatrix.m_32);
            m_localMatrix.m_43 = z + (v.m_x * m_localMatrix.m_13) + (v.m_y * m_localMatrix.m_23) + (v.m_z * m_localMatrix.m_33);

        } else if (relative2->m_parent == nullptr) { // relative to root axes
            Vector1x4 d = relative2->mapPos(Vector1x4(v.m_x, v.m_y, v.m_z, 0.f), m_parent.get());
            m_localMatrix.m_41 += d.m_x;
            m_localMatrix.m_42 += d.m_y;
            m_localMatrix.m_43 += d.m_z;

        } else if (relative2 == m_parent.get()) { // relative to parent axes
            m_localMatrix.m_41 += v.m_x;
            m_localMatrix.m_42 += v.m_y;
            m_localMatrix.m_43 += v.m_z;

        } else { // relative to arbitrary axes
            assert(m_parent.get());
            Vector1x4 d = relative2->mapPos(Vector1x4(v.m_x, v.m_y, v.m_z, 0.f), m_parent.get());
            m_localMatrix.m_41 += d.m_x;
            m_localMatrix.m_42 += d.m_y;
            m_localMatrix.m_43 += d.m_z;
        }
        invalidateSubtree();
    }

    void RefFrame::rotateX(float theta, const RefFrame* relative2) {
        if (relative2 == this || relative2 == nullptr) { // relative to own axes
            m_localMatrix = Matrix4x4(Matrix4x4::ROTATE_X, theta) * m_localMatrix;

        } else if (relative2 == m_parent.get()) { // relative to parent
            m_localMatrix = m_localMatrix * Matrix4x4(Matrix4x4::ROTATE_X, theta);

        } else { // relative to arbitrary axes
            assert(false);
        }
        invalidateSubtree();
    }

    void RefFrame::rotateY(float theta, const RefFrame* relative2) {
        if (relative2 == this || relative2 == nullptr) { // relative to own axes
            m_localMatrix = Matrix4x4(Matrix4x4::ROTATE_Y, theta) * m_localMatrix;

        } else if (relative2 == m_parent.get()) { // relative to parent
            m_localMatrix = m_localMatrix * Matrix4x4(Matrix4x4::ROTATE_Y, theta);

        } else { // relative to arbitrary axes
            assert(false);
        }
        invalidateSubtree();
    }

    void RefFrame::rotateZ(float theta, const RefFrame* relative2) {
        if (relative2 == this || relative2 == nullptr) { // relative to own axes
            m_localMatrix = Matrix4x4(Matrix4x4::ROTATE_Z, theta) * m_localMatrix;

        } else if (relative2 == m_parent.get()) { // relative to parent
            m_localMatrix = m_localMatrix * Matrix4x4(Matrix4x4::ROTATE_Z, theta);

        } else { // relative to arbitrary axes
            assert(false);
        }
        invalidateSubtree();
    }

    Vector1x4 RefFrame::mapPos(const Vector1x4& v, const RefFrame* tgt) const {
        if (tgt == m_parent.get()) {
            return v * m_localMatrix;
        } else if (tgt->m_parent == nullptr) {
            return v * getModelMatrix();
        } else {
            return v * getModelMatrix() * !tgt->getModelMatrix();
        }
    }

    json RefFrame::toJSON() const {
        const Matrix4x4& m = getModelMatrix();

        json node = {
            { "nodeType", "RefFrame" },
            { "modelMatrix", { m.m_11, m.m_12, m.m_13, m.m_14, m.m_21, m.m_22, m.m_23, m.m_24, m.m_31, m.m_32, m.m_33, m.m_34, m.m_41, m.m_42, m.m_43, m.m_44 } },
        };

        for (auto p = getChild(); p != nullptr; p = p->getNext()) {
            node["children"].push_back(p->toJSON());
        }

        return node;
    }

}
